### PR TITLE
PVR Enforce EDL limits and increase them

### PR DESCRIPTION
**Pull request by emveepee**
## Description

The c-api  between the PVR client  and core uses a hard-coded table to store an EDL array.  This PR ensures that clients do not overflow the array

The EDL table is returned as a vector from the client.   The API calls GetRecordingEdl() and GetEpgTagEdl()  include the size of the EDL arrays that are  being filled.  The code change happens If the passed size is less then the vector size the vector is resized to the passed size.

I am unsure how to test GetEpgTagEdl()  so it was excluded.

In addition the limit was extended to allow for longer running shows such as during the Olympics and Wimbledon.

## Motivation and context

Kodi will crash if the internal table overflows

Described on the forum https://forum.kodi.tv/showthread.php?tid=363766
## How has this been tested?

Created a test EDL file with 33 entries for a recording.  Kodi crashed without  the fix.  After the fix Kodi does not crash and the warning is logged.

Tested with current Nexus build on Windows x64 using pvr.nextpvr addon

Note because the change is in PVR.h the actual fix will be in the addon so impacted PVR clients will need to be rebuilt.

## What is the effect on users?

Kodi will not crash for recordings with lots of EDL entries.  Some entries will be lost if the maximum size is reached. A second commit could increase the limit for the current max 32 to another value, 60 has been proposed.

## Screenshots (if appropriate):

```
2021-07-30 21:44:43.483 T:18832 WARNING <general>: AddOnLog: pvr.nextpvr: CInstancePVRClient::ADDON_GetRecordingEdl: Truncating 33 EDL entries from client to permitted size 32
```

## Types of change

- [x ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by ksooo** [comment]
@AlwinEsch  does this require an add-on API version bump and/or recompile if every PVR add-on?

**Comment by emveepee** [comment]
> 
> 
> @AlwinEsch does this require an add-on API version bump and/or recompile of every PVR add-on?

@phunkyfish told me it does https://github.com/kodi-pvr/pvr.nextpvr/pull/175#issuecomment-890165895 but he would take care of that. for every addon that had this capability In testing changes, I had to recompile both.

**Comment by phunkyfish** [comment]
I was thinking of merging this after addons are branched for nexus and I’m bringing in the pvr providers work.

@emveepee could the same protection be added to GetEPGTagEdl?

It would make sense to have it for both.

**Comment by emveepee** [comment]
> @emveepee could the same protection be added to GetEPGTagEdl? It would make sense to have it for both. I have a way to test that here with a local contrived setup.

Sure, updated with the format diff.  Here is a sample edl with 33 breaks for your testing.  Did you want me to bump the 32 break limit too?  32 should be OK for regular use, except maybe profanity filter users.

Here is a test edl with 33 breaks [sample.zip](https://github.com/xbmc/xbmc/files/6911304/sample.zip) for your testing.

**Comment by phunkyfish** [comment]
> > @emveepee could the same protection be added to GetEPGTagEdl? It would make sense to have it for both. I have a way to test that here with a local contrived setup.
> 
> Sure, updated with the format diff. Did you want me to bump the 32 break limit too? 32 should be OK for regular use, except maybe profanity filter users.
> 
> Here is a test edl with 33 breaks [sample.zip](https://github.com/xbmc/xbmc/files/6911304/sample.zip) for your testing.

As this is only used when EDL is required I don't see the harm in setting it to 64. But let's see how likely it is to occur. For the user who hit this limit and caused a crash, how many EDL entries did they have?

**Comment by emveepee** [comment]
For historical perspective, in Leia size was passed to the addon and pvr.hts did respect it https://github.com/kodi-pvr/pvr.hts/blob/Leia/src/Tvheadend.cpp#L592

**Comment by phunkyfish** [comment]
> For historical perspective, in Leia size was passed to the addon and pvr.hts did respect it https://github.com/kodi-pvr/pvr.hts/blob/Leia/src/Tvheadend.cpp#L592

Yes, I would say that it not being handled correctly internally once we moved the API to C++ was an oversight.

**Comment by emveepee** [comment]
> As this is only used when EDL is required I don't see the harm in setting it to 64. But let's see how likely it is to occur. For the user who hit this limit and caused a crash, how many EDL entries did they have?

He provided the paste for the response from NextPVR (I first thought it could be backend issue) https://paste.kodi.tv/jezivezadu and there are 34 entries in what looks like a 5 hour recording.  A profanity mute EDL for an episode of Deadwould would need about 100 for one hour.

**Comment by phunkyfish** [comment]
I think long recordings for sports events like the olympics etc. Is reasonable. 64 sounds fine.

I’m not sure profanity skipping is that common a use case. I would imagine that for something like deadwood it would be very hard to follow with all the skips. WDYT?

**Comment by emveepee** [comment]
> I’m not sure profanity skipping is that common a use case. I would imagine that for something like deadwood it would be very hard to follow with all the skips. WDYT?

64 is fine, commit added, might as well make the .   Profanity EDL's use the mute option not skip. Here is a 66 item sample.
[sample.zip](https://github.com/xbmc/xbmc/files/6911844/sample.zip)

```
2021-07-31 17:21:16.930 T:18920 WARNING <general>: AddOnLog: pvr.nextpvr: CInstancePVRClient::ADDON_GetRecordingEdl: Truncating 66 EDL entries from client to permitted size 6
```

**Comment by phunkyfish** [comment]
@emveepee you can ignore the Jenkins diff above. We don’t tend to follow it for files this which only contain defines.

**Review by AlwinEsch — COMMENTED** [review]
From overview the changes looking good and more safe to in case addon brings more as the 64.

**Comment by phunkyfish** [comment]
> > Not for this. Any plug-in can generate an ID to represent a URL which is of reasonable length. This is even facilitated for plug-ins specifically by iptvsimple.
> 
> True and thus, the part of this PR increasing the size of the array from 32 to 64 is not actually needed. Just saying.

Fair, I updated the PR description to indicate that is for shows such as Wimbledon and The Olympics.

**Review by phunkyfish — APPROVED** [review]
This should be ready to go in before PVR Providers API 8.0.0 PR: https://github.com/xbmc/xbmc/pull/19395
